torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -176,19 +176,43 @@ template class UserRRef<py::object>;
 
 template <typename T>
 const T& OwnerRRef<T>::getValue() const {
-  // TODO: use callback to make this non-blocking
   std::unique_lock<std::mutex> lock(mutex_);
   valueCV_.wait(lock, [this] { return value_.has_value(); });
   return value_.value();
 }
 
 template <typename T>
-void OwnerRRef<T>::setValue(T&& value) {
-  {
-    std::lock_guard<std::mutex> lock(mutex_);
-    value_ = std::move(value);
+bool OwnerRRef<T>::hasValue() const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  return value_.has_value();
+}
+
+template <typename T>
+std::shared_ptr<FutureMessage> OwnerRRef<T>::getFuture() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (future_.get()) {
+    return future_;
+  }
+  future_ = std::make_shared<FutureMessage>();
+  std::shared_ptr<FutureMessage> ret = future_;
+  if (value_.has_value()) {
+    lock.unlock();
+    ret->markCompleted();
   }
+  return ret;
+}
+
+template <typename T>
+void OwnerRRef<T>::setValue(T&& value) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  value_ = std::move(value);
+  std::shared_ptr<FutureMessage> future;
+  future.swap(future_);
+  lock.unlock();
   valueCV_.notify_all();
+  if (future.get()) {
+    future->markCompleted();
```

Comment:
The same future can be mark completed twice in the following scenario? Is this intentional?

1. call `setValue`, so that `value_.has_value() == true`
2. call `getFuture`, a new `future_` will be created and mark completed.
3. call `setValue` again, `future_` will be swapped to `future` and call `markCompleted()` the second time. 

This would fail under current `FutureMessage` implementation. If this is not intentional, let's only mark complete when `future.complete() == false`.

Reply:
Thanks for bringing this up - I didn't consider the case where setValue() might be called twice.
I'll guard against this.